### PR TITLE
Minor refactor of reference entities to add support for reference launch plans in dynamic tasks

### PR DESCRIPTION
**Pull request by jeevb**
Signed-off-by: Jeev B <jeev.balakrishnan@freenome.com>

# TL;DR
Minor refactoring of how reference entities are handled with more specific handling of unsupported cases for `ReferenceWorkflow` and `ReferenceTask` specifically. This allows `ReferenceLaunchPlan` to work within dynamic tasks.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

**Comment by jeevb** [comment]
Smoke tested with:
```
import os
from typing import List

from flytekit import dynamic, workflow
from flytekit.core.launch_plan import reference_launch_plan
from flytekit.core.task import reference_task

PROJECT = os.environ["FLYTE_INTERNAL_PROJECT"]
DOMAIN = os.environ["FLYTE_INTERNAL_DOMAIN"]
WORKFLOW_NAME = "workflows.single_step.compute_square_wf"
TASK_NAME = "workflows.single_step.squared"
# Update this with a previously registered version. This simulates referencing
# launch plans registered in a different project/domain, with a different image.
VERSION = "0541ffc0dd80"


@dynamic
def dynamic_reference_compute_square_lp(input_integer: int) -> int:
    @reference_launch_plan(
        project=PROJECT, domain=DOMAIN, name=WORKFLOW_NAME, version=VERSION,
    )
    def _reference_compute_square_lp(input_integer: int) -> int:
        ...

    return _reference_compute_square_lp(input_integer=input_integer)


@reference_launch_plan(
    project=PROJECT, domain=DOMAIN, name=WORKFLOW_NAME, version=VERSION,
)
def reference_compute_square_lp(input_integer: int) -> int:
    ...


@reference_task(
    project=PROJECT, domain=DOMAIN, name=TASK_NAME, version=VERSION,
)
def reference_compute_square_task(value: int) -> int:
    ...


@workflow
def reference_compute_square_wf_wrapper(input_integer: int) -> List[int]:
    a = dynamic_reference_compute_square_lp(input_integer=input_integer)
    b = reference_compute_square_lp(input_integer=input_integer)
    c = reference_compute_square_task(value=input_integer)
    return [a, b, c]
```

**Comment by jeevb** [comment]
The codecov report is confusing. Not sure why it would drop dramatically in the unrelated files.

**Review by wild-endeavor — APPROVED** [review]
